### DISCUSSION TITLE
Multi GPU calculations

### DISCUSSION THREAD

**Discussion by LLapsus** [netket/netket#992] [General]
Hello, 
I run NetKet 3.1.2 on a computational node with 4 GPUs. When I check jax.devices(), I can see all 4 GPUs.
However, during the calculation NetKet is using just one GPU.
Is there a way how to utilize all the GPUs in parallel?
Thanks in advance.

**Comment by PhilipVinc** [comment]
You have to use MPI and make sure that every MPI rank sees a different device as default device.

If you want more specific instructions I'll try to cobble a script together to make this more clear

**Reply by LLapsus** [reply]
Thank you. That would be great.
I am using MPI, but when I run the script using mpirun, I can see just the script running on multiple CPUs, but it is still using just one GPU.

**Reply by PhilipVinc** [reply]
Be careful that you have to set the visible device in the environment variable *BEFORE* loading jax (and netket, as netket loads jax)

```python
import os

# detect MPI rank
from mpi4py import MPI
rank = MPI.COMM_WORLD.Get_rank()

# set only one visible device
os.environ["CUDA_VISIBLE_DEVICES"] = f"{rank}"
# force to use gpu
os.environ["JAX_PLATFORM_NAME"] = "gpu"

import jax 
import netket as nk

print(f"{rank} -> {jax.devices()}")

# 1D Lattice
L = 20
g = nk.graph.Hypercube(length=L, n_dim=1, pbc=True)
...
```

**Reply by joshuazlin** [reply]
Can I ask if this is still the recommended way to utilise multiple GPUs? Calling this python script with a slurm script

```bash
#!/bin/bash
#SBATCH --partition=devel
#SBATCH --time=00:30:00
#SBATCH --nodes=1
#SBATCH --tasks=2
#SBATCH --gres=gpu:2
mpirun -n 2 python -m netket.tools.check_mpi
mpirun -n 2 python test.py
```
is not working for me, it gives me the output

```
mpi4py_available             : True
mpi4jax_available            : True
available_cpus (rank 0)      : 2
n_nodes                      : 2
mpi4py | MPI version         : (3, 1)
mpi4py | MPI library_version : Open MPI v4.1.4, package: Open MPI root@cluster Distribution, ident: 4.1.4, repo rev: v4.1.4, May 26, 2022

1 -> [StreamExecutorGpuDevice(id=0, process_index=0, slice_index=0), StreamExecutorGpuDevice(id=1, process_index=0, slice_index=0)]
0 -> [StreamExecutorGpuDevice(id=0, process_index=0, slice_index=0), StreamExecutorGpuDevice(id=1, process_index=0, slice_index=0)]

and ssh-ing into the node, I see that both processes are still seeing both GPUs, 

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A     64533      C   python                            156MiB |
|    0   N/A  N/A     64534      C   python                            156MiB |
|    1   N/A  N/A     64533      C   python                            156MiB |
|    1   N/A  N/A     64534      C   python                            156MiB |
+-----------------------------------------------------------------------------+
```
I've run out of ideas of what to check/fix unfortunately

**Reply by PhilipVinc** [reply]
Hi @joshuazlin 
The snippet I posted works well for us, so it's hard for me to pinpoint exactly what is working and what is not working...

I suspect some recent changes in jax might be the root cause, and maybe the locally-set environment variables are not working correctly...

There are two things you might do. First: if you are using slurm try changing the batch script to the following:

```bash
#!/bin/bash
#SBATCH --partition=devel
#SBATCH --time=00:30:00
#SBATCH --nodes=1
#SBATCH --tasks=2
#SBATCH --gres=gpu:2

export CUDA_VISIBLE_DEVICES="$(echo $CUDA_VISIBLE_DEVICES | cut -d ',' -f $(($SLURM_LOCALID+1)))"

mpirun -n 2 python -m netket.tools.check_mpi
mpirun -n 2 python test.py
```

if this works, great.
If this does not work you can try setting the default device

**Comment by LLapsus** [comment]
@PhilipVinc Thank you. It works well :)
